test/utils/resource.go

```diff
@@ -0,0 +1,93 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package utils
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"regexp"
+
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/klog/v2"
+	ctrlClient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// CreateK8sResources creates all kubernetes resources within the given list.
+func CreateK8sResources(ctx context.Context, objects []ctrlClient.Object, cli ctrlClient.Client, testNamespace string) (map[string][]ctrlClient.ObjectKey, error) {
+	klog.Info("Creating K8s resources.")
+
+	objectKeys := make(map[string][]ctrlClient.ObjectKey)
+	for _, obj := range objects {
+
+		// Set a different namespace for each test case for isolation.
+		obj.SetNamespace(testNamespace)
+
+		gvk := obj.GetObjectKind().GroupVersionKind().String()
+		objectKeys[gvk] = append(objectKeys[gvk], ctrlClient.ObjectKeyFromObject(obj))
+		if err := cli.Create(ctx, obj); err != nil {
+			return nil, fmt.Errorf("failed to create %s %s/%s: %v", gvk, obj.GetNamespace(), obj.GetName(), err)
+		}
+		klog.Infof("Created %s %s/%s", gvk, obj.GetNamespace(), obj.GetName())
+	}
+	return objectKeys, nil
+}
+
+// DeleteK8sResources deletes all kubernetes resources within the given list.
+func DeleteK8sResources(ctx context.Context, objects []ctrlClient.Object, cli ctrlClient.Client) error {
+	klog.Info("Deleting K8s resources.")
+	for _, obj := range objects {
+		gvk := obj.GetObjectKind().GroupVersionKind().String()
+		err := cli.Delete(ctx, obj, &ctrlClient.DeleteOptions{})
+		if err != nil {
+			klog.Infof("Failed to delete %s %s/%s: %v", gvk, obj.GetNamespace(), obj.GetName(), err)
```

Comment:
This assumes that the obj is the same as the one in Create so it has the right namespace.

Reply:
Updated. See https://github.com/GoogleCloudPlatform/gke-networking-recipes/pull/143#discussion_r1290362505.